File: src/workflow/levels/level3_standard.py
"""
L3 標準模式 (Standard)

完整會話、驗證步驟。
適用於: 多章節開發、角色塑造、功能開發。
"""

from typing import Any, Dict, List, Optional
from dataclasses import dataclass
from .base_level import BaseLevel, LevelRegistry
from ..base_state import BaseState
from ...agents.base import AgentType


@dataclass
class PlanPhase:
    """計劃階段"""
    phase: int
    name: str
    description: str
    status: str = "pending"  # pending | in_progress | completed
    output: Optional[str] = None


@LevelRegistry.register(3)
class Level3Standard(BaseLevel):
    """
    L3 標準模式

    特點:
    - 完整 5 階段計劃
    - Architect → Writer → Critic 循環
    - 會話持久化
    - 驗證步驟

    命令鏈:
    plan → plan-verify → execute → verify
    """

    level = 3
    name = "standard"
    description = "標準模式 - 完整會話、驗證步驟"

    PLAN_PHASES = [
        PlanPhase(1, "需求理解", "理解用戶需求，澄清模糊點"),
        PlanPhase(2, "內容分析", "分析已有章節、角色、世界觀"),
        PlanPhase(3, "變更範圍", "定義修改範圍，識別影響點"),
        PlanPhase(4, "實施計劃", "生成詳細執行步驟"),
        PlanPhase(5, "驗證定義", "定義驗收標準和測試用例"),
    ]

    def __init__(
        self,
        architect: Any = None,
        writer: Any = None,
        critic: Any = None,
        config: Dict = None,
        container = None
    ):
        """
        初始化 L3 標準模式

        Args:
            architect: ArchitectAgent 實例（可選，優先使用 DI）
            writer: WriterAgent 實例（可選，優先使用 DI）
            critic: CriticAgent 實例（可選，優先使用 DI）
            config: 配置字典
            container: ServiceContainer 實例（依賴注入）
        """
        super().__init__(config, container)
        self._architect = architect
        self._writer = writer
        self._critic = critic

    def _get_architect(self):
        """獲取 Architect Agent（懶加載，通過 DI）"""
        if self._architect is None:
            self._architect = self.container.get_agent(AgentType.ARCHITECT)
        return self._architect

    def _get_writer(self):
        """獲取 Writer Agent（懶加載，通過 DI）"""
        if self._writer is None:
            self._writer = self.container.get_agent(AgentType.WRITER)
        return self._writer

    def _get_critic(self):
        """獲取 Critic Agent（懶加載，通過 DI）"""
        if self._critic is None:
            self._critic = self.container.get_agent(AgentType.CRITIC)
        return self._critic

    def execute(self, state: BaseState, **kwargs) -> BaseState:
        """
        執行標準工作流

        流程:
        1. Plan: 生成 5 階段計劃
        2. Plan-Verify: 驗證計劃完整性
        3. Execute: 調用 Writer 執行
        4. Critic: 評估質量
        5. 循環直到通過或達到 max_revisions
        """
        config = self.get_default_config()
        config.update(self.config or {})

        max_revisions = config.get("max_revisions", 3)
        pass_score = config.get("pass_score", 80)  # L3 执行策略 fallback，非 novel 发布门槛

        # Phase 1: 計劃
        state = self._plan_phase(state)

        # Phase 2: 計劃驗證
        if not self._verify_plan(state):
            state["errors"] = state.get("errors", []) + ["計劃驗證失敗"]
            state["decision"] = "HUMAN_REVIEW"
            return state

        # Phase 3-5: 執行循環
        revision_count = 0
        while revision_count < max_revisions:
            # 執行
            state = self._execute_phase(state)

            # 評估
            state = self._critic_phase(state)

            score = state.get("score", 0)
            decision = state.get("decision", "REVISE")

            if decision == "APPROVED" or score >= pass_score:
                state["decision"] = "APPROVED"
                break

            if decision == "HUMAN_REVIEW":
                break

            revision_count += 1
            state["revision_count"] = revision_count

        if revision_count >= max_revisions and state.get("decision") != "APPROVED":
            state["decision"] = "HUMAN_REVIEW"
            state["requires_human_intervention"] = True

        return state

    def get_required_agents(self) -> List[str]:
        return ["architect", "writer", "critic"]

    def plan(self, state: BaseState) -> List[PlanPhase]:
        """生成計劃 (5 階段)"""
        return self._plan_phase(state).get("plan_phases", self.PLAN_PHASES)

    def plan_verify(self, state: BaseState) -> Dict:
        """驗證計劃完整性"""
        return {
            "valid": self._verify_plan(state),
            "missing": [],
            "warnings": [],
        }

    def get_default_config(self) -> Dict:
        """获取默认配置（L3 执行策略阈值；非 novel 发布门槛）"""
        return {
            "max_revisions": 3,
            "pass_score": 80,
            "verbose": True,
            "retrieval_profile": "standard_balanced",
        }

    def _plan_phase(self, state: BaseState) -> BaseState:
        """計劃階段"""
        try:
            architect = self._get_architect()
            result = architect.run({
                "user_request": state.get("user_request", ""),
                "context": state.get("context", ""),
                "mode": "planning",
            })

            state["plan_phases"] = result.get("phases", self.PLAN_PHASES)
            state["implementation_plan"] = result.get("plan", {})

        except Exception as e:
            state["errors"] = state.get("errors", []) + [f"計劃失敗: {e}"]
            state["plan_phases"] = self.PLAN_PHASES

        return state

    def _verify_plan(self, state: BaseState) -> bool:
        """驗證計劃"""
        plan = state.get("implementation_plan", {})
        phases = state.get("plan_phases", [])

        # 基本檢查
        if not plan or not phases:
            return False

        # 檢查所有階段是否有輸出
        for phase in phases:
            if isinstance(phase, dict) and not phase.get("output"):
                return False

        return True

    def _execute_phase(self, state: BaseState) -> BaseState:
        """執行階段"""
        try:
            writer = self._get_writer()
            result = writer.run({
                "plan": state.get("implementation_plan", {}),
                "context": state.get("context", ""),
                "feedback": state.get("feedback_context", ""),
                "mode": "execute",
            })

            state["draft_content"] = result.get("content", "")
            state["draft_version"] = state.get("draft_version", 0) + 1

        except Exception as e:
            state["errors"] = state.get("errors", []) + [f"執行失敗: {e}"]

        return state

    def _critic_phase(self, state: BaseState) -> BaseState:
        """評估階段"""
        try:
            critic = self._get_critic()
            result = critic.run({
                "content": state.get("draft_content", ""),
                "plan": state.get("implementation_plan", {}),
                "mode": "evaluate",
            })

            state["score"] = result.get("score", 0)
            state["decision"] = result.get("decision", "REVISE")
            state["feedback_context"] = result.get("feedback", "")

        except Exception as e:
            state["errors"] = state.get("errors", []) + [f"評估失敗: {e}"]
            state["decision"] = "HUMAN_REVIEW"

        return state

File: src-ts/workflow/levels/level3-standard.ts
/**
 * level3-standard.ts - L3 Standard mode
 *
 * Full session, verification steps.
 * Suitable for: multi-chapter development, character development, feature development.
 *
 * Command chain:
 * plan -> plan-verify -> execute -> verify
 *
 * Migrated from src/workflow/levels/level3_standard.py
 */

import type { BaseState } from '../state.js';
import { AgentType } from '../../agents/base.js';
import type { IServiceContainer } from './level1-rapid.js';

// ============================================================
// Data types
// ============================================================

export interface PlanPhase {
  phase: number;
  name: string;
  description: string;
  status: 'pending' | 'in_progress' | 'completed';
  output: string | null;
}

// ============================================================
// Level3Standard
// ============================================================

export class Level3Standard {
  readonly level = 3;
  readonly name = 'standard';
  readonly description = '標準模式 - 完整會話、驗證步驟';

  static readonly PLAN_PHASES: PlanPhase[] = [
    { phase: 1, name: '需求理解', description: '理解用戶需求，澄清模糊點', status: 'pending', output: null },
    { phase: 2, name: '內容分析', description: '分析已有章節、角色、世界觀', status: 'pending', output: null },
    { phase: 3, name: '變更範圍', description: '定義修改範圍，識別影響點', status: 'pending', output: null },
    { phase: 4, name: '實施計劃', description: '生成詳細執行步驟', status: 'pending', output: null },
    { phase: 5, name: '驗證定義', description: '定義驗收標準和測試用例', status: 'pending', output: null },
  ];

  protected config: Record<string, unknown>;
  protected _container: IServiceContainer | null;
  private _architect: { run(input: Record<string, unknown>): Record<string, unknown> } | null;
  private _writer: { run(input: Record<string, unknown>): Record<string, unknown> } | null;
  private _critic: { run(input: Record<string, unknown>): Record<string, unknown> } | null;

  constructor(
    options?: {
      config?: Record<string, unknown>;
      container?: IServiceContainer;
      architect?: { run(input: Record<string, unknown>): Record<string, unknown> };
      writer?: { run(input: Record<string, unknown>): Record<string, unknown> };
      critic?: { run(input: Record<string, unknown>): Record<string, unknown> };
    },
  ) {
    this.config = options?.config ?? {};
    this._container = options?.container ?? null;
    this._architect = options?.architect ?? null;
    this._writer = options?.writer ?? null;
    this._critic = options?.critic ?? null;
  }

  protected get container(): IServiceContainer {
    if (this._container == null) {
      throw new Error('ServiceContainer not available.');
    }
    return this._container;
  }

  private _getArchitect(): { run(input: Record<string, unknown>): Record<string, unknown> } {
    if (this._architect == null) {
      this._architect = this.container.getAgent(AgentType.ARCHITECT);
    }
    return this._architect;
  }

  private _getWriter(): { run(input: Record<string, unknown>): Record<string, unknown> } {
    if (this._writer == null) {
      this._writer = this.container.getAgent(AgentType.WRITER);
    }
    return this._writer;
  }

  private _getCritic(): { run(input: Record<string, unknown>): Record<string, unknown> } {
    if (this._critic == null) {
      this._critic = this.container.getAgent(AgentType.CRITIC);
    }
    return this._critic;
  }

  /**
   * Execute the standard workflow
   *
   * Flow:
   * 1. Plan: generate 5-phase plan
   * 2. Plan-Verify: verify plan completeness
   * 3. Execute: call Writer
   * 4. Critic: evaluate quality
   * 5. Loop until passed or max_revisions reached
   */
  execute(state: BaseState): BaseState {
    const config = { ...this.getDefaultConfig(), ...this.config };

    const maxRevisions = (config.max_revisions as number) ?? 3;
    const passScore = (config.pass_score as number) ?? 80; // L3 execution policy fallback, not novel publish threshold

    // Phase 1: plan
    state = this._planPhase(state);

    // Phase 2: plan verification
    if (!this._verifyPlan(state)) {
      state.errors = [...(state.errors ?? []), '計劃驗證失敗'];
      state.decision = 'HUMAN_REVIEW';
      return state;
    }

    // Phase 3-5: execution loop
    let revisionCount = 0;
    while (revisionCount < maxRevisions) {
      state = this._executePhase(state);
      state = this._criticPhase(state);

      const score = state.score ?? 0;
      const decision = state.decision ?? 'REVISE';

      if (decision === 'APPROVED' || score >= passScore) {
        state.decision = 'APPROVED';
        break;
      }

      if (decision === 'HUMAN_REVIEW') {
        break;
      }

      revisionCount++;
      state.revision_count = revisionCount;
    }

    if (revisionCount >= maxRevisions && state.decision !== 'APPROVED') {
      state.decision = 'HUMAN_REVIEW';
      state.requires_human_intervention = true;
    }

    return state;
  }

  getRequiredAgents(): string[] {
    return ['architect', 'writer', 'critic'];
  }

  getDefaultConfig(): Record<string, unknown> {
    return {
      max_revisions: 3,
      pass_score: 80,
      verbose: true,
      retrieval_profile: 'standard_balanced',
    };
  }

  // ========== Public API ==========

  plan(state: BaseState): PlanPhase[] {
    const updated = this._planPhase(state);
    return ((updated as Record<string, unknown>).plan_phases as PlanPhase[]) ?? Level3Standard.PLAN_PHASES;
  }

  planVerify(state: BaseState): { valid: boolean; missing: string[]; warnings: string[] } {
    return {
      valid: this._verifyPlan(state),
      missing: [],
      warnings: [],
    };
  }

  // ========== Internal phases ==========

  _planPhase(state: BaseState): BaseState {
    try {
      const architect = this._getArchitect();
      const result = architect.run({
        user_request: state.user_request ?? '',
        context: state.context ?? '',
        mode: 'planning',
      });

      (state as Record<string, unknown>).plan_phases = result.phases ?? Level3Standard.PLAN_PHASES;
      (state as Record<string, unknown>).implementation_plan = result.plan ?? {};
    } catch (e) {
      const errorMsg = e instanceof Error ? e.message : String(e);
      state.errors = [...(state.errors ?? []), `計劃失敗: ${errorMsg}`];
      (state as Record<string, unknown>).plan_phases = Level3Standard.PLAN_PHASES;
    }

    return state;
  }

  _verifyPlan(state: BaseState): boolean {
    const plan = (state as Record<string, unknown>).implementation_plan;
    const phases = (state as Record<string, unknown>).plan_phases;

    if (!plan || !phases) return false;

    if (Array.isArray(phases)) {
      for (const phase of phases) {
        if (typeof phase === 'object' && phase !== null && !(phase as Record<string, unknown>).output) {
          return false;
        }
      }
    }

    return true;
  }

  _executePhase(state: BaseState): BaseState {
    try {
      const writer = this._getWriter();
      const result = writer.run({
        plan: (state as Record<string, unknown>).implementation_plan ?? {},
        context: state.context ?? '',
        feedback: ((state as Record<string, unknown>).feedback_context as string) ?? '',
        mode: 'execute',
      });

      state.draft_content = (result.content as string) ?? '';
      const draftVersion = ((state as Record<string, unknown>).draft_version as number) ?? 0;
      (state as Record<string, unknown>).draft_version = draftVersion + 1;
    } catch (e) {
      const errorMsg = e instanceof Error ? e.message : String(e);
      state.errors = [...(state.errors ?? []), `執行失敗: ${errorMsg}`];
    }

    return state;
  }

  _criticPhase(state: BaseState): BaseState {
    try {
      const critic = this._getCritic();
      const result = critic.run({
        content: state.draft_content ?? '',
        plan: (state as Record<string, unknown>).implementation_plan ?? {},
        mode: 'evaluate',
      });

      state.score = (result.score as number) ?? 0;
      state.decision = (result.decision as string) ?? 'REVISE';
      (state as Record<string, unknown>).feedback_context = (result.feedback as string) ?? '';
    } catch (e) {
      const errorMsg = e instanceof Error ? e.message : String(e);
      state.errors = [...(state.errors ?? []), `評估失敗: ${errorMsg}`];
      state.decision = 'HUMAN_REVIEW';
    }

    return state;
  }
}
